server/src/controllers/grades.ts

```diff
@@ -354,3 +357,322 @@ export async function addGrades(req: Request, res: Response, next: NextFunction)
+    // model type.
+  }) as unknown as Array<AttainmentInfo>;
+
+  /*
+   * Then we need to find the formulas used to calculate the grade of each
+   * attainment.
+   *
+   * For each attainment, we will construct a formula node object containing:
+   *   - The formula implementation storing the actual formula function as well
+   *     as the Yup schema for validating its parameters.
+   *   - The formula nodes of the subattainments of this attainment.
+   *   - The formula parameters needed for calculating the grade of this
+   *     attainment's parent attainment.
+   */
+
+  /*
+   * Stores the formula nodes of each attainment.
+   * Attainment ID -> Formula node of the given attainment.
+   */
+  const formulaNodesByAttainmentId: Map<number, FormulaNode> = new Map();
+
+  /*
+   * First find the formula implementation and formula parameters for the parent
+   * attainment's formula.
+   */
+  for (const attainment of attainments) {
+    const formula: Formula | null = attainment.formula;
+
+    // Ensure that the formula specified for this attainment is valid.
+    if (!yup.string().oneOf(Object.values(Formula)).required().validate(formula)) {
+      throw new ApiError(
+        `invalid formula ${formula} for attainment with ID ${attainment.id}`,
+        HttpCode.InternalServerError
+      );
+    }
+
+    formulaNodesByAttainmentId.set(attainment.id, {
+      formulaImplementation: await getFormulaImplementation(formula as Formula),
+      subFormulaNodes: [],
+      parentFormulaParams: attainment.parentFormulaParams
+    });
+  }
+
+  /*
+   * Then we will find the formula nodes of each attainment's subattainments.
+   * This will construct a tree structure, and a reference to the root of this
+   * tree is stored in the rootFormulaNode variable.
+   */
+  let rootFormulaNode: FormulaNode | null = null;
+
+  /*
+   * Local utility function to get a formula node from formulaNodesByAttainmentId
+   * as type FormulaNode, or throw an error if the formula node is undefined.
+   */
+  function getFormulaNode(attainmentId: number): FormulaNode {
+    const formulaNode: FormulaNode | undefined = formulaNodesByAttainmentId.get(attainmentId);
+
+    if (!formulaNode) {
+      throw new ApiError(
+        `found undefined formula node for attainment with ID ${attainmentId}`,
+        HttpCode.InternalServerError
+      );
+    }
+
+    return formulaNode;
+  }
+
+  for (const attainment of attainments) {
+    const formulaNode: FormulaNode = getFormulaNode(attainment.id);
+
+    /*
+     * Check whether this attainment is the root attainment. The root attainment
+     * has no parent.
+     */
+    if (attainment.parentId === null) {
+      /*
+       * This attainment is the root attainment. Store a reference to its formula
+       * node in the rootFormulaNode variable.
+       */
+
+      // There should only be one root attainment.
+      if (rootFormulaNode) {
+        throw new ApiError(
+          'duplicate root attainment',
```

Comment:
What if there are two separate attainment "trees" on the instance, one for example with exam attainments and other one with A+ attainments? Then there will be 2 attainments that don't have parents but only children, and this error is raised.

Reply:
Ok, so the exam and other possible attainment trees are stored under the one root attainment. 

Good point on the final exam vs. course exam and other points. If I remember correctly same thing was briefly discussed on one meeting last week with teaching staff.

But as this is the first iteration of grade calculation, I think we can ignore this for now and update later if needed.

Reply:
Actually even with final exams, we wouldn't need two root attainments. We could still have a single root attainment for the final grade with a formula such as "choose best," and under that, there are two attainments, one for the final exam and the other for regular course completion.

But this should be discussed in more detail later.